### PR TITLE
Da/poke

### PR DESCRIPTION
**Pull request by tadad**
adding new (better) interface(s) for I/O apart from the dojo

### CODE REVIEW

**Comment by tadad** [comment]
tested
`%task` with `:pyro &action [%note ~nec %i [%request [%'GET' 'https://google.com' ~ ~] *outbound-config:iris]]`
and
`%poke` with `:pyro &action [%poke ~nec ~nec %time %noun ~]`

**Comment by hosted-fornet** [comment]
Trying to test like:

```hoon
:pyro|dojo ~nec ":wallet +dbug"

:pyro &action [%poke ~nec ~nec %uqbar %wallet-poke [%transaction from=0x7a9a.97e0.ca10.8e1e.273f.0000.8dca.2b04.fc15.9f70 contract=0x74.6361.7274.6e6f.632d.7367.697a town=0x0 action=[%give to=0xd6dc.c8ff.7ec5.4416.6d4e.b701.d1a6.8e97.b464.76de amount=123.456 item=0x89a0.89d8.dddf.d13a.418c.0d93.d4b4.e7c7.637a.d56c.96c0.7f91.3a14.8174.c7a7.71e6]]]

:pyro|dojo ~nec ":wallet +dbug"

:pyro &action [%poke ~nec ~nec %uqbar %wallet-poke [%submit from=0x7a9a.97e0.ca10.8e1e.273f.0000.8dca.2b04.fc15.9f70 hash=[yourhash] gas=[rate=1 bud=1.000.000]]]

:pyro|dojo ~nec ":wallet +dbug"

:pyro &action [%dojo ~nec ":sequencer|batch"]

:pyro|dojo ~nec ":wallet +dbug"
```

but after the first command, the wallet is not properly populated with a pending transaction and I get this ouput:

```hoon
[ who=~nec
  %unknown-effect
  wire=/
    card
  [ p=%unto
      q
    [ 7.738.135.432.355.606.384
      0
      [1.717.658.988 103 97 108 108 58 32 114 97 119 45 112 111 107 101 32 118 97 108 101 32 102 97 105 108 32 58 117 113 98 97 114 32 37 119 97 108 108 101 116 45 112 111 107 101 0]
      [1.702.063.986 [[58 0] 0 0] [1.702.063.986 [[47 0] [47 0] 0] [1.717.658.988 115 117 114 0] [1.717.658.988 122 105 103 0] [1.717.658.988 119 97 108 108 101 116 0] [1.717.658.988 104 111 111 110 0] 0] [1.717.658.988 60 91 54 32 49 51 93 46 91 54 32 51 54 93 62 0] 0]
      [1.702.063.986 [[58 0] 0 0] [1.702.063.986 [[47 0] [47 0] 0] [1.717.658.988 115 117 114 0] [1.717.658.988 122 105 103 0] [1.717.658.988 119 97 108 108 101 116 0] [1.717.658.988 104 111 111 110 0] 0] [1.717.658.988 60 91 49 51 55 32 49 49 93 46 91 49 51 55 32 49 56 93 62 0] 0]
      [1.702.063.986 [[58 0] 0 0] [1.702.063.986 [[47 0] [47 0] 0] [1.717.658.988 115 117 114 0] [1.717.658.988 122 105 103 0] [1.717.658.988 119 97 108 108 101 116 0] [1.717.658.988 104 111 111 110 0] 0] [1.717.658.988 60 91 49 51 55 32 49 50 93 46 91 49 51 55 32 49 56 93 62 0] 0]
      [1.702.063.986 [[58 0] 0 0] [1.702.063.986 [[47 0] [47 0] 0] [1.717.658.988 115 117 114 0] [1.717.658.988 122 105 103 0] [1.717.658.988 119 97 108 108 101 116 0] [1.717.658.988 104 111 111 110 0] 0] [1.717.658.988 60 91 49 51 54 32 55 93 46 91 49 52 50 32 57 93 62 0] 0]
      [1.702.063.986 [[58 0] 0 0] [1.702.063.986 [[47 0] [47 0] 0] [1.717.658.988 115 117 114 0] [1.717.658.988 122 105 103 0] [1.717.658.988 119 97 108 108 101 116 0] [1.717.658.988 104 111 111 110 0] 0] [1.717.658.988 60 91 49 51 54 32 55 93 46 91 49 52 50 32 57 93 62 0] 0]
      [1.702.063.986 [[58 0] 0 0] [1.702.063.986 [[47 0] [47 0] 0] [1.717.658.988 115 117 114 0] [1.717.658.988 122 105 103 0] [1.717.658.988 119 97 108 108 101 116 0] [1.717.658.988 104 111 111 110 0] 0] [1.717.658.988 60 91 49 48 50 32 51 93 46 91 49 52 51 32 53 93 62 0] 0]
      [1.702.063.986 [[58 0] 0 0] [1.702.063.986 [[47 0] [47 0] 0] [1.717.658.988 115 121 115 0] [1.717.658.988 118 97 110 101 0] [1.717.658.988 103 97 108 108 0] [1.717.658.988 104 111 111 110 0] 0] [1.717.658.988 60 91 55 51 51 32 50 55 93 46 91 55 51 51 32 52 56 93 62 0] 0]
      0
    ]
  ]
]
```

What is going on here, am I doing something wrong?

**Comment by hosted-fornet** [comment]
> Trying to test like:
> 
> ```hoon
> :pyro|dojo ~nec ":wallet +dbug"
> 
> :pyro &action [%poke ~nec ~nec %uqbar %wallet-poke [%transaction from=0x7a9a.97e0.ca10.8e1e.273f.0000.8dca.2b04.fc15.9f70 contract=0x74.6361.7274.6e6f.632d.7367.697a town=0x0 action=[%give to=0xd6dc.c8ff.7ec5.4416.6d4e.b701.d1a6.8e97.b464.76de amount=123.456 item=0x89a0.89d8.dddf.d13a.418c.0d93.d4b4.e7c7.637a.d56c.96c0.7f91.3a14.8174.c7a7.71e6]]]
> 
> :pyro|dojo ~nec ":wallet +dbug"
> 
> :pyro &action [%poke ~nec ~nec %uqbar %wallet-poke [%submit from=0x7a9a.97e0.ca10.8e1e.273f.0000.8dca.2b04.fc15.9f70 hash=[yourhash] gas=[rate=1 bud=1.000.000]]]
> 
> :pyro|dojo ~nec ":wallet +dbug"
> 
> :pyro &action [%dojo ~nec ":sequencer|batch"]
> 
> :pyro|dojo ~nec ":wallet +dbug"
> ```
> 
> but after the first command, the wallet is not properly populated with a pending transaction and I get this ouput:
> 
> ```hoon
> [ who=~nec
>   %unknown-effect
>   wire=/
>     card
>   [ p=%unto
>       q
>     [ 7.738.135.432.355.606.384
>       0
>       [1.717.658.988 103 97 108 108 58 32 114 97 119 45 112 111 107 101 32 118 97 108 101 32 102 97 105 108 32 58 117 113 98 97 114 32 37 119 97 108 108 101 116 45 112 111 107 101 0]
>       [1.702.063.986 [[58 0] 0 0] [1.702.063.986 [[47 0] [47 0] 0] [1.717.658.988 115 117 114 0] [1.717.658.988 122 105 103 0] [1.717.658.988 119 97 108 108 101 116 0] [1.717.658.988 104 111 111 110 0] 0] [1.717.658.988 60 91 54 32 49 51 93 46 91 54 32 51 54 93 62 0] 0]
>       [1.702.063.986 [[58 0] 0 0] [1.702.063.986 [[47 0] [47 0] 0] [1.717.658.988 115 117 114 0] [1.717.658.988 122 105 103 0] [1.717.658.988 119 97 108 108 101 116 0] [1.717.658.988 104 111 111 110 0] 0] [1.717.658.988 60 91 49 51 55 32 49 49 93 46 91 49 51 55 32 49 56 93 62 0] 0]
>       [1.702.063.986 [[58 0] 0 0] [1.702.063.986 [[47 0] [47 0] 0] [1.717.658.988 115 117 114 0] [1.717.658.988 122 105 103 0] [1.717.658.988 119 97 108 108 101 116 0] [1.717.658.988 104 111 111 110 0] 0] [1.717.658.988 60 91 49 51 55 32 49 50 93 46 91 49 51 55 32 49 56 93 62 0] 0]
>       [1.702.063.986 [[58 0] 0 0] [1.702.063.986 [[47 0] [47 0] 0] [1.717.658.988 115 117 114 0] [1.717.658.988 122 105 103 0] [1.717.658.988 119 97 108 108 101 116 0] [1.717.658.988 104 111 111 110 0] 0] [1.717.658.988 60 91 49 51 54 32 55 93 46 91 49 52 50 32 57 93 62 0] 0]
>       [1.702.063.986 [[58 0] 0 0] [1.702.063.986 [[47 0] [47 0] 0] [1.717.658.988 115 117 114 0] [1.717.658.988 122 105 103 0] [1.717.658.988 119 97 108 108 101 116 0] [1.717.658.988 104 111 111 110 0] 0] [1.717.658.988 60 91 49 51 54 32 55 93 46 91 49 52 50 32 57 93 62 0] 0]
>       [1.702.063.986 [[58 0] 0 0] [1.702.063.986 [[47 0] [47 0] 0] [1.717.658.988 115 117 114 0] [1.717.658.988 122 105 103 0] [1.717.658.988 119 97 108 108 101 116 0] [1.717.658.988 104 111 111 110 0] 0] [1.717.658.988 60 91 49 48 50 32 51 93 46 91 49 52 51 32 53 93 62 0] 0]
>       [1.702.063.986 [[58 0] 0 0] [1.702.063.986 [[47 0] [47 0] 0] [1.717.658.988 115 121 115 0] [1.717.658.988 118 97 110 101 0] [1.717.658.988 103 97 108 108 0] [1.717.658.988 104 111 111 110 0] 0] [1.717.658.988 60 91 55 51 51 32 50 55 93 46 91 55 51 51 32 52 56 93 62 0] 0]
>       0
>     ]
>   ]
> ]
> ```
> 
> What is going on here, am I doing something wrong?

Note the second step here will require `[yourhash]` be filled in with the tx hash from the previous step -- but the first step seems to not work.

**Comment by hosted-fornet** [comment]
To be more explicit what I did was install then

```hoon
:ziggurat &ziggurat-action [%foo %new-project ~]
:ziggurat &ziggurat-action [%foo %start-pyro-ships ~[~nec ~bud]]
:ziggurat &ziggurat-action [%foo %add-and-queue-test `%setup /zig/test-steps/setup/hoon]
:ziggurat &ziggurat-action [%$ %run-queue ~]

:pyro|dojo ~nec ":wallet +dbug"

:pyro &action [%poke ~nec ~nec %uqbar %wallet-poke [%transaction from=0x7a9a.97e0.ca10.8e1e.273f.0000.8dca.2b04.fc15.9f70 contract=0x74.6361.7274.6e6f.632d.7367.697a town=0x0 action=[%give to=0xd6dc.c8ff.7ec5.4416.6d4e.b701.d1a6.8e97.b464.76de amount=123.456 item=0x89a0.89d8.dddf.d13a.418c.0d93.d4b4.e7c7.637a.d56c.96c0.7f91.3a14.8174.c7a7.71e6]]]

:pyro|dojo ~nec ":wallet +dbug"
```

where each wallet +dbug is checking to make sure we are in proper state.

**Comment by tadad** [comment]
@hosted-fornet there are some issues I'm figuring out but I think the `%transaction` isn't formatted correctly above. I tried to do : 
`^-  wallet-poke:w  [%transaction from=0x7a9a.97e0.ca10.8e1e.273f.0000.8dca.2b04.fc15.9f70 contract=0x74.6361.7274.6e6f.632d.7367.697a town=0x0 action=[%give to=0xd6dc.c8ff.7ec5.4416.6d4e.b701.d1a6.8e97.b464.76de amount=123.456 item=0x89a0.89d8.dddf.d13a.418c.0d93.d4b4.e7c7.637a.d56c.96c0.7f91.3a14.8174.c7a7.71e6]]`

**Comment by tadad** [comment]
the only issue I've found so far is that sometimes it will force you to create a mark file, even when you are creating a dumb-mark e.g. `%rollup-action` in `/app/rollup/hoon`

**Comment by hosted-fornet** [comment]
> @hosted-fornet there are some issues I'm figuring out but I think the `%transaction` isn't formatted correctly above. I tried to do : `^- wallet-poke:w [%transaction from=0x7a9a.97e0.ca10.8e1e.273f.0000.8dca.2b04.fc15.9f70 contract=0x74.6361.7274.6e6f.632d.7367.697a town=0x0 action=[%give to=0xd6dc.c8ff.7ec5.4416.6d4e.b701.d1a6.8e97.b464.76de amount=123.456 item=0x89a0.89d8.dddf.d13a.418c.0d93.d4b4.e7c7.637a.d56c.96c0.7f91.3a14.8174.c7a7.71e6]]`

Shoot, you're right. There is an origin field now; we need to update the readme. My bad

**Comment by hosted-fornet** [comment]
Looks like adding a ~ after %transaction works as expected. Doing a bit more testing then will finish review.

File: app/pyro.hoon [hunk 1]
```diff
@@ -311,10 +311,10 @@
...
       ..abet-pe
     =.  ..abet-pe
-      =/  sof  ((soft unix-effect) i.effects)
-      ?~  sof
-        ~?  aqua-debug=&  [who=who %unknown-effect i.effects]
-        ..abet-pe
+      ?~  sof=((soft unix-effect) i.effects)
+        ?:  &(=(+<:i.effects %unto) ?=(^ +>+.i.effects))
+          ((slog (flop ;;(tang +>+>.i.effects))) ~&(who=who ..abet-pe))
+        ~&([who=who %unknown-effect i.effects] ..abet-pe)
```
[the remark this answers is not in the document]
Comment (tadad) [reply app/pyro.hoon:317]: just fixed. For some reason I thought it was all untyped when I wrote this

**Pull request merged**